src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,182 @@
+            Assert.Throws<Exception>(() => df.Select(udf(df["NameIdsMap"])).Show());
+
+            // Currently, there is a workaround to support MapType using Hashtable.
+            Func<Column, Column> workingUdf = Udf<Hashtable, string>(
+                dict => dict.Count.ToString());
+
+            Row[] rows = df.Select(workingUdf(df["NameIdsMap"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+
+            var expected = new[] { "1", "1", "1" };
+            string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();
+            Assert.Equal(expected, rowsToArray);
+        }
+
+        /// <summary>
+        /// UDF that returns Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnAsMapType()
+        {
+            // UDF with return as MapType throws a following exception:
+            // Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
+            // ---> System.NotImplementedException: The method or operation is not implemented.
+            // System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in MicrTosoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                str => new Dictionary<string, string> { { str, str } });
+            Assert.Throws<NotImplementedException>(
+                () => _df.Select(udf(_df["name"])).Collect().ToArray());
+
+            //Show() works here. See the example below.
+            _df.Select(udf(_df["name"])).Show();
+        }
+
+        /// <summary>
+        /// UDF that takes in Row type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithRowType()
+        {
+            Func<Column, Column> udf = Udf<Row, string>(
+                (row) =>
+                {
+                    string city = row.GetAs<string>("city");
+                    string state = row.GetAs<string>("state");
+                    return $"{city},{state}";
+                });
+
+            Row[] rows = _df.Select(udf(_df["info"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+
+            var expected = new[] { "Burdwan,Paschimbanga", "Los Angeles,California", "Seattle," };
+            string[] rowsToArray = rows.Select(x => x[0].ToString()).ToArray();

```

Comment:
let's do `actual` to go along with `expected`.

Reply:
I have changed all the above comments accordingly. Thanks!